src/vm/types/mod.rs

```diff
@@ -0,0 +1,450 @@
+            })
+    }
+}
+
+impl PartialEq for ListData {
+    fn eq(&self, other: &ListData) -> bool {
+        self.data == other.data
+    }
+}
+
+impl Hash for ListData {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.data.hash(state);
+    }
+}
+
+impl PartialEq for TupleData {
+    fn eq(&self, other: &TupleData) -> bool {
+        self.data_map == other.data_map
+    }
+}
+
+impl Hash for TupleData {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.data_map.hash(state);
+    }
+}
+
+pub const NONE: Value = Value::Optional(OptionalData { data: None });
+
+impl Value {
+    pub fn some(data: Value) -> Value {
+        Value::Optional(OptionalData {
+            data: Some(Box::new(data)) })
+    }
+
+    pub fn none() -> Value {
+        Value::Optional(OptionalData {
+            data: None })
+    }
+
+    pub fn okay(data: Value) -> Value {
+        Value::Response(ResponseData { 
+            committed: true,
+            data: Box::new(data) })
+    }
+
+    pub fn error(data: Value) -> Value {
+        Value::Response(ResponseData { 
+            committed: false,
+            data: Box::new(data) })
+    }
+
+    /// Invariant: the supplied Values have already been "checked", i.e., it's a valid Value object
+    ///  this invariant is enforced through the Value constructors, each of which checks to ensure
+    ///  that any typing data is correct.
+    pub fn list_with_type(list_data: Vec<Value>, expected_type: ListTypeData) -> Result<Value> {
+        if expected_type.size()? > MAX_VALUE_SIZE {
+            return Err(RuntimeErrorType::ValueTooLarge.into())
+        }
+
+        if (expected_type.max_len as usize) < list_data.len() {
+            return Err(InterpreterError::FailureConstructingListWithType.into())
+        }
+
+        let expected_item_type = expected_type.get_list_item_type();
+
+        for item in &list_data {
+            if !expected_item_type.admits(&item) {
+                return Err(InterpreterError::FailureConstructingListWithType.into())
+            }
+        }
+
+        Ok(Value::List(ListData { data: list_data, type_signature: expected_type }))
+    }
+
+    pub fn list_from(list_data: Vec<Value>) -> Result<Value> {
+        let type_sig = TypeSignature::construct_parent_list_type(&list_data)?;
+        // Aaron: at this point, we've _already_ allocated memory for this type.
+        //     (e.g., from a (map...) call, or a (list...) call.
+        //     this is a problem _if_ the static analyzer cannot already prevent
+        //     this case. This applies to all the constructor size checks.
+        if type_sig.size()? > MAX_VALUE_SIZE {
+            return Err(RuntimeErrorType::ValueTooLarge.into())
+        }
+        Ok(Value::List(ListData { data: list_data, type_signature: type_sig }))
+    }
+
+    pub fn buff_from(buff_data: Vec<u8>) -> Result<Value> {
+        if buff_data.len() > u32::max_value() as usize {
```

Comment:
Is this check necessary, given that `MAX_VALUE_SIZE` is smaller than 4 billion bytes?  Is there a material difference between `BufferTooLarge` and `ValueTooLarge`?

Reply:
Nope -- though in some indeterminate future when it MAX_VALUE_SIZE may be way too large, this check will be necessary. Also, no, there's no reason for both BufferTooLarge and ValueTooLarge to exist. BufferTooLarge is eliminated in the value-bounding branch. (For what it's worth, this code isn't really new in this PR -- it's just appearing as new because of breakings types.rs into mod.rs, signatures.rs, and serialization.rs)